indimail-x/incrypt.1
.TH incrypt 1

.SH NAME
incrypt - command line implementation of crypt(3)

.SH SYNOPSYS
\fBincrypt\fR [ -h hash ] [ -S salt ] passphrase
.I salt

.SH DESCRIPTION
incrypt is a command line impementation of crypt(3) password encryption
function. The crypt functions irreversibly “hash” phrase for storage in the
system password database (shadow(5)) using a cryptographic “hashing
method.” The result of this operation is called a “hashed passphrase” or
just a “hash.” Hashing methods are described in crypt(5).

\fIhash\fR controls which hashing method to use, and also supplies various
parameters to the chosen method, most importantly a random “salt” which en‐
sures that no two stored hashes are the same, even if the phrase strings
are the same.

If
.I salt
is a character string starting with the characters "$\fIid\fP$"
followed by a string terminated by "$":
.RS

$\fIid\fP$\fIsalt\fP$\fIencrypted\fP

.RE
then instead of using the DES machine,
.I id
identifies the encryption method used and this then determines how the rest
of the password string is interpreted.
The following values of
.I id
are supported:
.RS
.TS
l l.
ID  | Method
_
1   | MD5
2a  | Blowfish (not in mainline glibc; added in some
    | Linux distributions)
.\" openSUSE has Blowfish, but AFAICS, this option is not supported
.\" natively by glibc -- mtk, Jul 08
.\"
.\" md5 | Sun MD5
.\" glibc doesn't appear to natively support Sun MD5; I don't know
.\" if any distros add the support.
5   | SHA-256 (since glibc 2.7)
6   | SHA-512 (since glibc 2.7)
.TE
.RE

So $5$\fIsalt\fP$\fIencrypted\fP is an SHA-256 encoded password and
$6$\fIsalt\fP$\fIencrypted\fP is an SHA-512 encoded one.

"\fIsalt\fP" stands for the up to 16 characters following "$\fIid\fP$" in
the salt. The encrypted part of the password string is the actual computed
password. The size of this string is fixed:
.TS
l l.
MD5     | 22 characters
SHA-256 | 43 characters
SHA-512 | 86 characters
.TE

The characters in "\fIsalt\fP" and "\fIencrypted\fP" are drawn from the set
[\fBa\-zA\-Z0\-9./\fP].
In the MD5 and SHA implementations the entire
.I key
is significant (instead of only the first
8 bytes in DES).

Warning: The key space consists of 2**56 equal 7.2e16 possible values.
Exhaustive searches of this key space are possible using massively parallel
computers. Software, such as crack(1), is available which will search the
portion of this key space that is generally used by humans for passwords.
Hence, password selection should, at minimum, avoid common words and names.
The use of a passwd(1) program that checks for crackable passwords during
the selection process is recommended.

.SH OPTIONS

.TP 3
\fB\-h\fR \fIhash\fR
Specifiy hash which is one of DES, MD5, SHA-256, SHA-512. Here \fIID\fR is
the id in \fI/etc/shadow\fR. See \fBincrypt\fR(1).
.TS
l l.
HASH    ID Description
_
DES     -  DES encryption (shouldn't be used)
MD5     1  MD5 encryption (shouldn't be used)
SHA-256 5  SHA256 encryption
SHA-512 6  SHA512 encryption
.TE

.TP 3
\fB\-S\fR \fIsalt\fR
Specify a base64 encoded salt to be used when generating SCRAM password. If
not specified, this will be generated using libsodium/gsasl. Here base64
implies characters [0-9], [a-z], [A-Z] and the two characters [./].

.SH NOTE on DES Encryption
The original hashing method from Unix V7, based on the DES block cipher.
Because DES is cheap on modern hardware, because there are only 4096
possible salts and 2**56 possible hashes, and because it truncates
passphrases to 8 characters, it is feasible to discover any passphrase
hashed with this method. It should only be used if you absolutely have to
generate hashes that will work on an old operating system that supports
nothing else.

It is based on the Data Encryption Standard algorithm with variations
intended (among other things) to discourage use of hardware implementations
of a key search.

\fIpassphrase\fR is a user's typed password.

\fIsalt\fR is a two-character string chosen from the set [a-zA-Z0-9./].
This string is used to perturb the algorithm in one of 4096 different
ways.

By taking the lowest 7 bits of each of the first eight characters of the
key, a 56-bit key is obtained. This 56-bit key is used to encrypt
repeatedly a constant string (usually a string consisting of all zeros).
The returned value points to the encrypted password, a series of 13
printable ASCII characters (the first two characters represent the salt
itself).  The return value points to static data whose content is
overwritten by each call.

.SH RETURN VALUES
On success, the encrypted password is printed on stdout. On error, non zero
status is returned.

.SH SEE ALSO
crypt(3)


indimail-x/NEWS
doc/RELEASE-Notes-indimail.md

indimail-x/incrypt.c
/*
 * $Log: incrypt.c,v $
 * Revision 1.5  2023-07-17 11:44:45+05:30  Cprogrammer
 * set hash method from hash_method control file in controldir
 *
 * Revision 1.4  2023-07-16 22:40:14+05:30  Cprogrammer
 * added YESCRYPT hash
 *
 * Revision 1.3  2023-01-22 10:40:03+05:30  Cprogrammer
 * replaced qprintf with subprintf
 *
 * Revision 1.2  2022-08-28 11:57:37+05:30  Cprogrammer
 * added option to specify salt and hash method
 *
 * Revision 1.1  2022-08-26 21:34:45+05:30  Cprogrammer
 * Initial revision
 *
 */
#ifdef HAVE_CONFIG_H
#include "config.h"
#endif
#ifndef _XOPEN_SOURCE
#define _XOPEN_SOURCE
#endif
#include <unistd.h>
#ifdef HAVE_CRYPT_H
#include <crypt.h>
#else
char           *crypt(const char *, const char *);
#endif
#ifdef HAVE_QMAIL
#include <strerr.h>
#include <substdio.h>
#include <subfd.h>
#include <str.h>
#include <fmt.h>
#include <in_crypt.h>
#include <valid_password_chars.h>
#include <sgetopt.h>
#include <hashmethods.h>
#include <makesalt.h>
#include <env.h>
#include <scan.h>
#endif
#include "common.h"
#include "get_hashmethod.h"

#define FATAL   "incrypt: fatal: "
#define WARN    "incrypt: warning: "

static char    *usage = "usage: incrypt [-h hash] -S salt passphrase";

int
main(int argc, char **argv)
{
	char           *ptr, *passphrase = (char *) NULL, *hash = (char *) NULL;
	char            strnum[FMT_ULONG], salt[SALTSIZE + 5];
	int             c, i, r, verbose;

	*salt = 0;
	verbose = 0;
	while ((c = getopt(argc, argv, "vh:S:")) != opteof) {
		switch (c)
		{
		case 'h':
			if ((r = scan_int(optarg, &i)) != str_len(optarg))
				i = -1;
			if (!str_diffn(optarg, "DES", 3) || i == DES_HASH)
				strnum[fmt_int(strnum, DES_HASH)] = 0;
			else
			if (!str_diffn(optarg, "MD5", 3) || i == MD5_HASH)
				strnum[fmt_int(strnum, MD5_HASH)] = 0;
			else
			if (!str_diffn(optarg, "SHA-256", 7) || i == SHA256_HASH)
				strnum[fmt_int(strnum, SHA256_HASH)] = 0;
			else
			if (!str_diffn(optarg, "SHA-512", 7) || i == SHA512_HASH)
				strnum[fmt_int(strnum, SHA512_HASH)] = 0;
			else
			if (!str_diffn(optarg, "YESCRYPT", 8) || i == YESCRYPT_HASH)
				strnum[fmt_int(strnum, YESCRYPT_HASH)] = 0;
			else {
				strerr_die5x(100, FATAL, "wrong hash method ", optarg,
						". Supported HASH Methods: DES MD5 SHA-256 SHA-512 YESCRYPT\n", usage);
			}
			hash = optarg;
			if (!env_put2("PASSWORD_HASH", strnum))
				strerr_die1x(111, "out of memory");
			break;
		case 'S':
			i = fmt_str(salt, optarg);
			salt[i] = 0;
			break;
		case 'v':
			verbose = 1;
			break;
		default:
			strerr_die2x(100, WARN, usage);
			break;
		}
	}
	if (!hash && !env_get("PASSWORD_HASH")) {
		if ((i = get_hashmethod((char *) 0)) == -1)
			strerr_die2sys(111, FATAL, "get_hashmethod: ");
		strnum[fmt_int(strnum, i)] = 0;
		if (!env_put2("PASSWORD_HASH", strnum))
			strerr_die1x(111, "out of memory");
	}
	if (!*salt) {
		if (makesalt(salt, SALTSIZE) == -1)
			strerr_die1sys(111, "failed to generate salt: ");
		if (verbose) {
			subprintfe(subfdout, "incrypt", "generated salt [%s]\n", salt);
			flush("incrypt");
		}
	} else {
		/* see crypt(5) for valid salt */
		if (salt[0] == '$' && !(salt[2] == '$' || salt[3] == '$' ||
				!str_diffn(salt, "$sha1", 5) ||
				!str_diffn(salt, "$2b$", 4) ||
				!str_diffn(salt, "$md5", 4)))
			strerr_die3x(100, FATAL, "invalid salt ", salt);
	}
	if (optind != (argc - 1))
		strerr_die2x(100, WARN, usage);
	passphrase = argv[optind++];
	if (!valid_password_chars(passphrase))
		strerr_die2x(100, FATAL, "invalid character used in passphrase");
	if (!(ptr = in_crypt(passphrase, salt)))
		strerr_die5x(100, FATAL, "failed to crypt passphrase ", passphrase, " with salt ", salt);
	subprintfe(subfdout, "incrypt", "\"%s\"\n", ptr);
	flush("incrypt");
	return(0);
}
